File: Engine/src/main/java/org/etieskrill/engine/util/FixedArrayDeque.java
package org.etieskrill.engine.util;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayDeque;
import java.util.Collection;

/**
 * An implementation of {@link ArrayDeque} with a fixed size. If any element is added while the maximum capacity is
 * reached, the first element that was added from the perspective of the writing direction will be overridden, which is
 * the element at {@code tail} for {@link FixedArrayDeque#addFirst(Object)}, and {@code head} for
 * {@link FixedArrayDeque#addLast(Object)}.
 *
 * @param <E> the type of elements in the collection
 */
public class FixedArrayDeque<E> extends ArrayDeque<E> {

    private final int size;

    public FixedArrayDeque(int numElements) {
        super(numElements);
        this.size = numElements;
    }

    @Override
    public void addFirst(@NotNull E e) {
        if (size() > 0 && size() >= size)
            super.removeLast();
        super.addFirst(e);
    }

    @Override
    public void addLast(@NotNull E e) {
        if (size() > 0 && size() >= size)
            super.removeFirst();
        super.addLast(e);
    }

    @Override
    public boolean addAll(Collection<? extends E> c) {
        for (E element : c) addFirst(element);
        return true;
    }
}

File: Engine/src/main/java/org/etieskrill/engine/util/FixedArrayDeque.kt
package org.etieskrill.engine.util;

/**
 * An implementation of [ArrayDeque] with a fixed size. If any element is added while the maximum capacity is reached,
 * the first element that was added from the perspective of the writing direction will be overridden, which is the
 * element at [tail](ArrayDeque.last) for [addFirst], and [head](ArrayDeque.first) for [addLast].
 *
 * @param E the type of elements in the collection
 */
class FixedArrayDeque<E>(numElements: Int) : Collection<E> {

    val capacity: Int = numElements
    override val size: Int get() = deque.size

    val deque = ArrayDeque<E>(numElements)

    val isFull get() = deque.size == capacity

    fun add(element: E) = addLast(element)

    fun addFirst(element: E) {
        if (deque.isNotEmpty() && deque.size >= capacity) {
            deque.removeLast()
        }
        deque.addFirst(element)
    }

    fun addLast(element: E) {
        if (deque.isNotEmpty() && deque.size >= capacity) {
            deque.removeFirst()
        }
        deque.addLast(element)
    }

    fun addAll(collection: Collection<E>) = collection.forEach(this::addLast)

    fun fill(element: E) = deque.fill(element)

    operator fun plusAssign(element: E) {
        deque.add(element)
    }

    override fun isEmpty() = deque.isEmpty()
    override fun contains(element: E) = deque.contains(element)
    override fun iterator() = deque.iterator()
    override fun containsAll(elements: Collection<E>) = deque.containsAll(elements)

}

@JvmName(name = "averageOfInt")
fun FixedArrayDeque<Int>.average() = deque.average()
@JvmName(name = "averageOfLong")
fun FixedArrayDeque<Long>.average() = deque.average()
@JvmName(name = "averageOfFloat")
fun FixedArrayDeque<Float>.average() = deque.average()
@JvmName(name = "averageOfDouble")
fun FixedArrayDeque<Double>.average() = deque.average()
